_posts/2016-10-25-cve-2016-5195-dirtycow-mitigated.md

```diff
@@ -0,0 +1,47 @@
+
+Because the `force` flag is only useful in very specific circumstances, only certain code paths can trigger the vulnerability. Kernel security engineer and Sandstorm contributor Andrew Lutomirski tells us the only entry points appear to be:
+
+* The `ptrace()` system call's `PTRACE_POKEDATA` operation, which is explicitly meant to be used by debuggers, often for the purpose of setting breakpoints.
+* Writes to `/proc/<pid>/mem`. This appears to be a mistake -- there's no apparent reason why this should use `force`. Probably, someone was copy-pasting code.
+* Various drivers, which are also probably using the flag by mistake.
+
+As it turns out, none of these code paths can be exploited by Sandstorm apps:
+
+* Sandstorm uses seccomp to block the app from invoking `ptrace()`.
+* Sandstorm does not mount `/proc` inside app sandboxes.
+* Sandstorm does not expose driver interfaces from inside app sandboxes. For example, `/dev` contains only `null`, `zero`, `urandom`, and `random` (which is actually a synonym for `urandom`, [as it](https://gist.github.com/tarcieri/6347417) [should be](http://www.2uo.de/myths-about-urandom/)).
+
+So, as far as we can tell, Sandstorm has never been vulnerable to this bug.
+
+### Defense in depth
+
+Even if Sandstorm were vulnerable, the exploit would have far reduced impact inside Sandstorm than in a typical Linux environment, because:
+
+* A Sandstorm app's filesystem consists of the contents of its own package. It cannot see the host system's files nor files belonging to other apps, hence it would not be able to memory-map them in order to modify them using this bug.
+* App packages cannot contain setuid binaries and, even if they could, apps would not be able to execute them, because Sandstorm sets the `NO_NEW_PRIVS` `prctl()` flag inside the sandbox.
+
+Because of these considerations, even if Sandstorm had not mitigated the bug outright, the impact of the bug would be that one instance of an app could attack other instances of the same app (running on the same machine in separate containers). Most apps do not allow a remote user to upload arbitrary code to execute and, hence, would only be vulnerable if the app author had included attack code in the app. On Sandstorm, we would consider this a vulnerability, because Sandstorm seeks to implement "confinement", meaning that an app should not be able to "phone home" to its author or speak to other copies of itself without permission.
+
+Interestingly, though, on most platforms, it's fully expected that instances of the same app can voluntarily collude with each other, e.g. by communicating with a central coordinator or by simply opening network connections to each other. And actually, Sandstorm does not yet fully implement confinement either -- we intend to, and an app has to go out of its way to break confinement today, but we have not yet prioritized systematically closing all holes.
```

Comment:
This paragraph is a little hard to read. I can try to suggest a rewording tomorrow.